filament/backend/src/vulkan/VulkanDriver.cpp

```diff
@@ -888,12 +895,52 @@ void VulkanDriver::beginRenderPass(Handle<HwRenderTarget> rth, const RenderPassP
         }
     }
 
-    const VkCommandBuffer cmdbuffer = mContext.commands->get().cmdbuffer;
     VulkanAttachment depth = rt->getSamples() == 1 ? rt->getDepth() : rt->getMsaaDepth();
+#ifdef VULKAN_DEBUG_REAL
+    if (depth.texture) {
+        depth.texture->print();
+    }
+#endif
 
-    VulkanDepthLayout initialDepthLayout = fromVkImageLayout(depth.getLayout());
-    VulkanDepthLayout renderPassDepthLayout = VulkanDepthLayout::ATTACHMENT;
-    VulkanDepthLayout finalDepthLayout = VulkanDepthLayout::ATTACHMENT;
+    bool samplingDepthAttachment = false;
+    const VkCommandBuffer cmdbuffer = mContext.commands->get().cmdbuffer;
+    for (uint8_t samplerGroupIdx = 0; samplerGroupIdx < Program::SAMPLER_BINDING_COUNT;
+            samplerGroupIdx++) {
+        VulkanSamplerGroup* vksb = mSamplerBindings[samplerGroupIdx];
+        if (!vksb) {
+            continue;
+        }
+        SamplerGroup* sb = vksb->sb.get();
+        for (size_t i = 0; i < sb->getSize(); i++) {
+            const SamplerDescriptor* boundSampler = sb->data() + i;
+            if (UTILS_LIKELY(boundSampler->t)) {
+                VulkanTexture* texture = handle_cast<VulkanTexture*>(boundSampler->t);
+                if (!any(texture->usage & TextureUsage::DEPTH_ATTACHMENT)) {
+                    continue;
+                }
+                samplingDepthAttachment
+                        = depth.texture && texture->getVkImage() == depth.texture->getVkImage();
+                if (texture->getPrimaryImageLayout() == VulkanLayout::DEPTH_SAMPLER) {
+                    continue;
+                }
+                VkImageSubresourceRange subresources{
+                        .aspectMask = VK_IMAGE_ASPECT_DEPTH_BIT,
+                        .baseMipLevel = 0,
+                        .levelCount = texture->levels,
+                        .baseArrayLayer = 0,
+                        .layerCount = texture->depth,
+                };
+                texture->transitionLayout(cmdbuffer, subresources, VulkanLayout::DEPTH_SAMPLER);
+                break;
+            }
+        }
+    }
+    // The state of the layout after the transition in the above block
+    VulkanLayout currentDepthLayout = depth.getLayout();
+    VulkanLayout const renderPassDepthLayout = samplingDepthAttachment
```

Comment:
comment to explain the difference between DEPTH_SAMPLER and DEPTH_ATTACHMENT in this context?

Reply:
above comment should cover this.